Update rocm recipes for rocm-5.0.0 release

srekolam opened the pull request:
This PR has updates for the rocm-5.0.0 release. I had verified the builds for 4.5.2 and 5.0.0 releases. Verified a few hip examples etc.
I will submit a new PR or add to this existing PR the changes for 5.0.2 release depending the comments that i receive and the release availability for 5.0.2 etc .
Also, few changes for llvm-amdgpu package for adding libcxx and libcxxabi to LLVM_ENABLE_RUNTIMES along with a new recipe for miopentensile is in progress.
The rocm-tensile recipe would require #29082 changes.

srekolam committed:
update rocm recipes for ROCm-5.0.0 release

srekolam committed:
revert address sanitizer changes

srekolam committed:
fix style errors

srekolam committed:
Merge branch 'develop' into rocm-5.0_rel_br1

cgmb commented:
In ROCm 5.0, rocFFT introduced a dependency on sqlite. It's downloaded via CMake's FetchContent at build-time and embedded in the library. IDK what the Spack policy is for that sort of thing is, but if you need to patch it, you might find parts of [this recent commit](https://github.com/ROCmSoftwarePlatform/rocFFT/commit/4bb38e6c132debc11e9b9a73512a45271bfb1ee9) handy as a reference.

UweSauter commented:
> In ROCm 5.0, rocFFT introduced a dependency on sqlite. It's downloaded via CMake's FetchContent at build-time and embedded in the library. IDK what the Spack policy is for that sort of thing is, but if you need to patch it, you might find parts of [this recent commit](https://github.com/ROCmSoftwarePlatform/rocFFT/commit/4bb38e6c132debc11e9b9a73512a45271bfb1ee9) handy as a reference.

Downloads during build time is a no-go for certain sites – AMD knows this and needs to resolve that.

@srekolam What about 5.0.1?

srekolam commented:
> > In ROCm 5.0, rocFFT introduced a dependency on sqlite. It's downloaded via CMake's FetchContent at build-time and embedded in the library. IDK what the Spack policy is for that sort of thing is, but if you need to patch it, you might find parts of [this recent commit](https://github.com/ROCmSoftwarePlatform/rocFFT/commit/4bb38e6c132debc11e9b9a73512a45271bfb1ee9) handy as a reference.
> 
> Downloads during build time is a no-go for certain sites – AMD knows this and needs to resolve that.
> 
> @srekolam What about 5.0.1?

The current plan is to update the rocm recipes with 5.0.2 release. there is no plan to support 5.0.1

srekolam committed:
Merge branch 'develop' into rocm-5.0_rel_br1

srekolam committed:
Merge branch 'develop' into rocm-5.0_rel_br1

srekolam committed:
update to the rocm recipes for 5.0.2 release

trws commented on var/spack/repos/builtin/packages/hip/package.py:
```diff
@@ -84,6 +90,8 @@ class Hip(CMakePackage):
             when='@{0}'.format(d_version)
         )
     for d_version, d_shasum in [
+        ('5.0.2',  '34decd84652268dde865f38e66f8fb4750a08c2457fea52ad962bced82a03e5e'),
```
Why are the rocclr and opencl resources downloaded both here and in the hip-rocclr package? Is there some reason this can't be a dependency on hip-rocclr?

trws commented on var/spack/repos/builtin/packages/llvm-amdgpu/package.py:
```diff
@@ -36,7 +38,7 @@ class LlvmAmdgpu(CMakePackage):
     variant('build_type', default='Release', values=("Release", "Debug", "RelWithDebInfo"), description='CMake build type')
     variant('rocm-device-libs', default=True, description='Build ROCm device libs as external LLVM project instead of a standalone spack package.')
-    variant('openmp', default=True, description='Enable OpenMP')
+    variant('openmp', default=False, description='Enable OpenMP')
```
Why?

srekolam replied:
In the rocm stack, rocm-openmp-extras recipe would provide the openmp libraries. Also, variant Openmp was suppose to be false from the beginning.

trws replied:
If this has moved to another package, deprecate the variant so it can be removed.

trws commented on var/spack/repos/builtin/packages/rocfft/package.py:
```diff
@@ -40,18 +42,29 @@ class Rocfft(CMakePackage):
     depends_on('cmake@3:', type='build')
+    depends_on('python@3:', type='build', when='@5.0.0:')
+    depends_on('sqlite@3.36.0', type='build', when='@5.0.0:')
+    resource(name='sqlite',
```
Why is this both a dep and a resource, forget to remove resource?

srekolam replied:
i corrected it.. initially the fetchcontent for sqlite was okay but looked like in some airgapped systems ,it will fail. So replaced it with the resource.

trws replied:
That's not what I'm asking.  There is a `depends_on(sqlite)` right above it, you can link to the one spack built from that dependency *instead of* using a resource and building it in the cmake.  This solution is slightly better than the original, but it still requires downloading and building sqlite twice for little benefit.

trws commented on var/spack/repos/builtin/packages/rocfft/Improve-compilation-by-using-sqlite-recipe-for-rocfft.patch:
```diff
@@ -0,0 +1,30 @@
+-)
+-FetchContent_MakeAvailable(sqlite_local)
+-add_library( sqlite3 STATIC ${sqlite_local_SOURCE_DIR}/sqlite3.c )
++add_library( sqlite3 STATIC ${sql_DIR}/sqlite3.c )
```
Why is this compiling another sqlite rather than using the dependency?

trws reviewed (changes requested):
There are some things left to do here.  At a minimum the sqlite issue should get finished, but I'd really like to know what the repeated fetching of rocclr and opencl sources is about as well.

srekolam committed:
fix review comments inside the rocfft recipe

cgmb commented:
By the way, if you find problems in the build systems for the ROCm math libraries when integrating them into Spack, please be sure to open an issue/ticket on the library. Even if Spack can patch around it, the developers working upstream need to be told about problems. Making mistakes and being corrected is part of how people learn. It's an important feedback loop, and I know it has sometimes been missed in the past.

I'd already told the rocFFT devs about the SQLite issue and worked with them to get an option to use find_package into a future version of the library. So, they are aware of this one already, but feel free to reach out to them if you need help backporting the fix.

srekolam committed:
update the rocfft recipe for sqlite dependency

srekolam commented:
> By the way, if you find problems in the build systems for the ROCm math libraries when integrating them into Spack, please be sure to open an issue/ticket on the library. Even if Spack can patch around it, the developers working upstream need to be told about problems. Making mistakes and being corrected is part of how people learn. It's an important feedback loop, and I know it has sometimes been missed in the past.
> 
> I'd already told the rocFFT devs about the SQLite issue and worked with them to get an option to use find_package into a future version of the library. So, they are aware of this one already, but feel free to reach out to them if you need help backporting the fix.

Thanks @cgmb . I have updated the rocfft recipe

trws commented:
Thanks @srekolam, we're down to one thing I'd really like to know now. Why does this pull rocclr and opencl sources as resources instead of using the dependence on hip-rocclr?

srekolam commented:
@trws , here is what i know. i can get more information if its not clear. There were major changes in HIP between 4.3.1 and 4.5.0,  HIP used to have sub directories like amdocl and rocclr till hip@4.3.1 . amdocl directory contained  opencl header files ..these files were exact copies of the files that were available from the repo - rocm-opencl . from hip@4.5.0 , rocclr sub directory contained the actual implementation of HIP for amd platform . This subdirectory is now renamed as hipamd (https://github.com/ROCm-Developer-Tools/HIPamd) .  amdocl is now removed from hip repo and the header file etc were referenced through the opencl. 
The build instructions for hip from 4.5.0 is documented as below ( https://github.com/ROCm-Developer-Tools/HIP/blob/develop/docs/markdown/hip_build.md) which i have done thru resources. 
Build HIP:
cd "$HIPAMD_DIR"
mkdir -p build; cd build
cmake -DHIP_COMMON_DIR=$HIP_DIR -DAMD_OPENCL_PATH=$OPENCL_DIR -DROCCLR_PATH=$ROCCLR_DIR -DCMAKE_PREFIX_PATH="<ROCM_PATH>/" -DCMAKE_INSTALL_PREFIX=$PWD/install ..
make -j$(nproc)
sudo make install

trws commented:
Ok, the question I'm trying to get to is whether hip can build with the `-DROCCLR_PATH` set to an installed copy rather than a source copy.  If not that's something we should follow up with AMD about because that's really unfortunate, but if that's the case then it is how it is and I'll approve.

srekolam commented:
@trws , i can verify it with the internal team on HIP and will confirm if HIP can refer to installed copy of rocclr than source copy but from the documentation link i mentioned , i see a git clone of rocclr.git and setting of the ROCclr_DIR="$(readlink -f ROCclr)" .

trws commented:
Yeah that sounds like it needs the source, that really needs to get fixed.  I'll send something to the packaging working group for the Coral2 COE about it, if you could follow up internally as well that would be great, but we need this in more than we need it fixed this instant.  Thanks!

cgmb commented:
> Yeah that sounds like it needs the source, that really needs to get fixed.

[The distro packagers aren't thrilled about it, either](https://lists.debian.org/debian-ai/2022/02/msg00101.html). @Maxzor struggled with this when packaging hipamd for Debian.

cgmb commented:
@trws, I tried to reach you by email but I haven't heard back. I just wanted to mention it here in case you responded and your message got blocked by the spam filters.

I was wondering how you feel about the opposite solution to the weird multi-resource HIP package build. Rather than downloading rocclr, hipamd and opencl as resources and combining them to build HIP, what if all four components came together in a single repo? It would just be one repo that contained several tightly-coupled components (akin to the LLVM repo and its subprojects).

trws commented:
That would be fine, if AMD wants to switch to a monorepo approach for those components that would solve that problem, although we'd have to re-work our packaging some to deal with it.
